pkg/verification/service/verification_service.go

```diff
@@ -0,0 +1,274 @@
+	if convErr != nil {
+		// We shouldn't get an error here, but if we do, we will set verification attempts to max allowed
+		// so that we at least now have a valid value, and let the workflow continue to the
+		// subsequent attempts check
+		log.Error(ctx, convErr, fmt.Sprintf("error converting annotation [%s] value [%s] to integer, on UserSignup: [%s]",
+			v1alpha1.UserVerificationAttemptsAnnotationKey,
+			signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey], signup.Name))
+		attemptsMade = s.config.GetVerificationAttemptsAllowed()
+		signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(attemptsMade)
+	}
+
+	// If the user has made more attempts than is allowed per day, return an error
+	if attemptsMade >= s.config.GetVerificationAttemptsAllowed() {
+		verificationErr = errors.NewTooManyRequestsError("too many verification attempts", "")
+	}
+
+	if verificationErr == nil {
+		// Parse the verification expiry timestamp
+		exp, parseErr := time.Parse(TimestampLayout, signup.Annotations[v1alpha1.UserVerificationExpiryAnnotationKey])
+		if parseErr != nil {
+			// If the verification expiry timestamp is corrupt or missing, then return an error
+			verificationErr = errors.NewInternalError(parseErr, "error parsing expiry timestamp")
+		} else if now.After(exp) {
+			// If it is now past the expiry timestamp for the verification code, return a 403 Forbidden error
+			verificationErr = errors.NewForbiddenError("expired", "verification code expired")
+		}
+	}
+
+	if verificationErr == nil {
+		if code != signup.Annotations[v1alpha1.UserSignupVerificationCodeAnnotationKey] {
+			// The code doesn't match
+			attemptsMade++
+			signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(attemptsMade)
+			verificationErr = errors.NewForbiddenError("invalid code", "the provided code is invalid")
+		}
+	}
+
+	if verificationErr == nil {
+		// If the code matches then set VerificationRequired to false, reset other verification annotations
+		signup.Spec.VerificationRequired = false
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationCodeAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserVerificationAttemptsAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationCounterAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationTimestampAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationInitTimestampAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserVerificationExpiryAnnotationKey)
+	} else {
+		log.Error(ctx, verificationErr, "error validating verification code")
+	}
+
+	// Update changes made to the UserSignup
+	_, updateErr := s.Services().SignupService().UpdateUserSignup(signup)
+	if updateErr != nil {
+		log.Error(ctx, updateErr, "error updating UserSignup")
+		verificationErr = errors.NewInternalError(updateErr, "error updating UserSignup")
```

Comment:
why do we re-assign the error at this point? it doesn't look like anything is done with verificationErr after it's re-assigned.

Reply:
Errors while updating `UserSignup` take priority, which is why it's re-assigned before being returned at the end of the function.

Reply:
i guess my point is, since this error takes precedence, why even bother re-assigning? may as well just return?